addons/binding/org.openhab.binding.wifiled/src/main/java/org/openhab/binding/wifiled/handler/FadingWiFiLEDDriver.java

```diff
@@ -143,49 +175,132 @@ private void changeState(final InternalLedState newState) throws IOException {
         }
     }
 
-    private void fadeToState(final InternalLedState newState) throws IOException {
-        if (!newState.equals(realTargetState)) {
-            keepFading = false;
-            realTargetState = newState;
-
-            executorService.schedule(() -> {
-                if (currentTargetState.equals(newState)) {
-                    return;
-                }
-
-                keepFading = true;
-
-                try (Socket socket = new Socket(host, port)) {
-                    logger.debug("Connected to '{}'", socket);
-
-                    socket.setSoTimeout(DEFAULT_SOCKET_TIMEOUT);
-
-                    try (DataOutputStream outputStream = new DataOutputStream(socket.getOutputStream())) {
-                        // ensure controller is on
-                        sendRaw(getBytesForPower(true), outputStream);
+    /**
+     * Runnable that takes care of fading of the LED's
+     */
+    final static class LEDFaderRunner implements Runnable {
+        protected Logger logger = LoggerFactory.getLogger(LEDFaderRunner.class);
+        protected final int steps;
+
+        // Keep track of LED states
+        protected int fadeStepper = 1;
+        protected InternalLedState currentFadeState;
+        protected InternalLedState toState;
+        protected InternalLedState fromState;
+
+        // After the last update keep the communication channel open for XXX ms
+        protected long keepCommOpenForMS = 1000;
+        // Keep track of when last communication happened
+        protected long lastComTime = 0;
+        // Function that will be called to update LED
+        protected final Function<InternalLedState, Boolean> ledSender;
+
+        protected final Lock lock = new ReentrantLock();
+
+        public LEDFaderRunner(InternalLedState fromState, InternalLedState toState, int steps, int keepCommOpenForMS, Function<InternalLedState, Boolean> ledSender) {
+            this.currentFadeState = fromState;
+            this.fromState = fromState;
+            this.toState = toState;
+            this.steps = steps;
+            this.ledSender = ledSender;
+            this.keepCommOpenForMS = keepCommOpenForMS;
```

Comment:
Do you want to able to set this dynamic or can it just use the same constant as the parent class?